flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinator.java

```diff
@@ -0,0 +1,322 @@
+              DataStatisticsEvent.create(checkpointId, globalDataStatistics, statisticsSerializer);
+          int parallelism = parallelism();
+          for (int i = 0; i < parallelism; ++i) {
+            subtaskGateways[i].sendEvent(dataStatisticsEvent);
+          }
+
+          return null;
+        },
+        String.format(
+            "Failed to send operator %s coordinator global data statistics for checkpoint %d",
+            operatorName, checkpointId));
+  }
+
+  @Override
+  public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exception {
+    runInCoordinatorThread(
+        () -> {
+          LOG.debug("Handling event from subtask {} of {}: {}", subtask, operatorName, event);
+          Preconditions.checkArgument(event instanceof DataStatisticsEvent);
+          handleDataStatisticRequest(subtask, ((DataStatisticsEvent<D, S>) event));
+        },
+        String.format("handling operator event %s from subtask %d", event.getClass(), subtask));
+  }
+
+  @Override
+  public void checkpointCoordinator(long checkpointId, CompletableFuture<byte[]> resultFuture) {
+    runInCoordinatorThread(
+        () -> {
+          LOG.debug(
+              "Snapshotting data statistics coordinator {} for checkpoint {}",
+              operatorName,
+              checkpointId);
+          resultFuture.complete(
+              DataStatisticsUtil.serializeAggregatedStatistics(
+                  completedStatistics, statisticsSerializer));
+        },
+        String.format("taking checkpoint %d", checkpointId));
+  }
+
+  @Override
+  public void notifyCheckpointComplete(long checkpointId) {}
+
+  @Override
+  public void resetToCheckpoint(long checkpointId, @Nullable byte[] checkpointData)
+      throws Exception {
+    Preconditions.checkState(
+        !started, "The coordinator %s can only be reset if it was not yet started", operatorName);
+
+    if (checkpointData == null) {
+      LOG.info(
+          "Data statistic coordinator {} has nothing to restore from checkpoint {}",
+          operatorName,
+          checkpointId);
+      return;
+    }
+
+    LOG.info(
+        "Restoring data statistic coordinator {} from checkpoint {}", operatorName, checkpointId);
+    completedStatistics =
+        DataStatisticsUtil.deserializeAggregatedStatistics(checkpointData, statisticsSerializer);
+  }
+
+  @Override
+  public void subtaskFailed(int subtask, @Nullable Throwable reason) {
+    runInCoordinatorThread(
+        () -> {
+          LOG.info(
+              "Unregistering gateway after failure for subtask {} of data statistic {}",
+              subtask,
+              operatorName);
+          Preconditions.checkState(
+              this.coordinatorThreadFactory.isCurrentThreadCoordinatorThread());
+          subtaskGateways[subtask] = null;
+        },
+        String.format("handling subtask %d failure", subtask));
+  }
+
+  @Override
+  public void subtaskReset(int subtask, long checkpointId) {
```

Comment:
the interface is unchanged. But the subtaskGateway in Flink 1.16 is `Map<Integer, SubtaskGateway>[] gateways` who map key is attempt num, while in Flink 1.15 we don't need to track the attemp num. Thus this is nothing we need to do in `subtaskRest`.